man/socklog-conf.8
.\" Automatically generated by Pandoc 3.1.11.1
.\"
.TH "socklog\-conf" "8" 
.SH NAME
socklog\-conf \- sets up a socklog(8) service.
.SH SYNOPSIS
\f[B]socklog\-conf\f[R] unix|inet|klog|ucspi\-tcp \f[I]acct\f[R]
\f[I]logacct\f[R]
.PD 0
.P
.PD
\f[B]socklog\-conf\f[R] notify \f[I]acct\f[R] \f[I]grp\f[R]
.SH DESCRIPTION
\f[B]socklog\-conf\f[R] creates \f[I]/etc/sv\f[R] if necessary and a
subdirectory that runs the \f[B]unix\f[R], \f[B]inet\f[R],
\f[B]ucspi\-tcp\f[R] or \f[B]notify\f[R] service (see below for each
service).
Except for the notify service, \f[B]socklog\-conf\f[R] also creates a
log directory with subdirectories.
.PP
\f[I]acct\f[R], \f[I]logacct\f[R] and \f[I]grp\f[R] must not contain any
special characters.
.SH UNIX SERVICE
\f[B]socklog\-conf\f[R] unix \f[I]acct\f[R] \f[I]logacct\f[R]
.PP
Running \f[B]socklog\-conf\f[R] with the 1st argument \f[I]unix\f[R],
\f[B]socklog\-conf\f[R] creates the service directory
\f[I]/etc/sv/socklog\-unix\f[R] and the log directory
\f[I]/var/log/socklog\f[R].
.PP
\f[B]socklog\-conf\f[R] arranges for \f[B]socklog\f[R](8) to run under
the uid and gid of \f[I]acct\f[R] and to listen for syslog messages on
\f[I]/dev/log\f[R].
.PP
It also creates automatically rotated log directories in
\f[I]/var/log/socklog\f[R].
The logs are owned by \f[I]logacct\f[R].
The corresponding \f[B]svlogd\f[R](8) process runs under the uid and gid
of \f[I]logacct\f[R].
.PP
You can run the service under \f[B]runsvdir\f[R](8) by creating a
symbolic link in the services directory:
.PP
ln \-s /etc/sv/socklog\-unix /service/
.SH INET SERVICE
\f[B]socklog\-conf\f[R] inet \f[I]acct\f[R] \f[I]logacct\f[R]
.PP
Running \f[B]socklog\-conf\f[R] with the 1st argument \f[I]inet\f[R],
\f[B]socklog\-conf\f[R] creates the service directory
\f[I]/etc/sv/socklog\-inet\f[R] and the log directory
\f[I]/var/log/socklog\-inet\f[R].
.PP
\f[B]socklog\-conf\f[R] arranges for \f[B]socklog\f[R](8) to run under
the uid and gid of \f[I]acct\f[R] and to listen for syslog messages on
the UDP socket \f[I]0.0.0.0:514\f[R].
.PP
It also creates automatically rotated log directories in
\f[I]/var/log/socklog\-inet\f[R].
The logs are owned by \f[I]logacct\f[R].
The corresponding \f[B]svlogd\f[R](8) process runs under the uid and gid
of \f[I]logacct\f[R].
.PP
You can run the service under \f[B]runsvdir\f[R](8) by creating a
symbolic link in the services directory:
.PP
ln \-s /etc/sv/socklog\-inet /service/
.SH KLOG SERVICE
\f[B]socklog\-conf\f[R] klog \f[I]acct\f[R] \f[I]logacct\f[R]
.PP
Running \f[B]socklog\-conf\f[R] with the 1st argument \f[I]klog\f[R],
\f[B]socklog\-conf\f[R] creates the service directory
\f[I]/etc/sv/socklog\-klog\f[R] and the log directory
\f[I]/var/log/socklog\-klog\f[R].
.PP
\f[B]socklog\-conf\f[R] arranges for \f[B]socklog\f[R](8) to run under
the uid and gid of \f[I]acct\f[R] and to read kernel messages from
\f[I]/proc/kmsg\f[R] on Linux, or \f[I]/dev/socklog\-klog\f[R] on BSD.
.PP
It also creates automatically rotated log directories in
\f[I]/var/log/socklog\-klog\f[R].
The logs are owned by \f[I]logacct\f[R].
The corresponding \f[B]svlogd\f[R](8) process runs under the uid and gid
of \f[I]logacct\f[R].
.PP
You can run the service under \f[B]runsvdir\f[R](8) by creating a
symbolic link in the services directory:
.PP
ln \-s /etc/sv/socklog\-klog /service/
.SH UCSPI\-TCP SERVICE
\f[B]socklog\-conf\f[R] ucspi\-tcp \f[I]acct\f[R] \f[I]logacct\f[R]
.PP
Running \f[B]socklog\-conf\f[R] with the 1st argument
\f[I]ucspi\-tcp\f[R], \f[B]socklog\-conf\f[R] creates the service
directory \f[I]/etc/sv/socklog\-ucspi\-tcp\f[R] and the log directory
\f[I]/var/log/socklog\-ucspi\-tcp\f[R].
.PP
\f[B]socklog\-conf\f[R] arranges for \f[B]tcpsvd\f[R](1) to run
\f[B]socklog\f[R](8) under the uid and gid of \f[I]acct\f[R] and to
listen on the TCP socket \f[I]0.0.0.0:10116\f[R].
.PP
It also creates automatically rotated log directories in
\f[I]/var/log/socklog\-ucspi\-tcp\f[R].
The logs are owned by \f[I]logacct\f[R].
The corresponding \f[B]svlogd\f[R](8) process runs under the uid and gid
of \f[I]logacct\f[R].
.PP
You can run the service under \f[B]runsvdir\f[R](8) by creating a
symbolic link in the services directory:
.PP
ln \-s /etc/sv/socklog\-ucspi\-tcp /service/
.SH NOTIFY SERVICE
\f[B]socklog\-conf\f[R] notify \f[I]acct\f[R] \f[I]grp\f[R]
.PP
Running \f[B]socklog\-conf\f[R] with the 1st argument \f[I]notify\f[R],
\f[B]socklog\-conf\f[R] creates the service directory
\f[I]/etc/sv/socklog\-notify\f[R].
.PP
\f[B]socklog\-conf\f[R] arranges for \f[B]uncat\f[R](1) to run under the
uid and gid of \f[I]acct\f[R] and to listen on the named pipe
\f[I]/var/log/socklog/.notify\f[R].
The named pipe will have mode 0620, the uid of \f[I]acct\f[R] and the
gid of \f[I]grp\f[R].
.PP
All uids running a log service that is configured to push log events to
the socklog\-notify service must be member of the group \f[I]grp\f[R].
.PP
You can run the service under \f[B]runsvdir\f[R](8) by creating a
symbolic link in the services directory:
.PP
ln \-s /etc/sv/socklog\-notify /service/
.SH SEE ALSO
socklog(8), svlogd(8), nc(1), tryto(1), uncat(1), socklog\-check(8),
tcpsvd(8), sv(8), runsv(8), runsvdir(8)
.PP
https://smarden.org/socklog/
.PD 0
.P
.PD
https://smarden.org/runit/
.SH AUTHOR
Gerrit Pape <pape\[at]smarden.org>


package/CHANGES
  * doc/benefits.html, doc/notify.html: use chpst -u instead of setuidgid.
  * doc/configuration.html, doc/notify.html, doc/upgrade.html,
    man/socklog-conf.8: use /service instead of /var/service.
  * doc/index.html: spelling.
  * doc/install.html: how to install as non-root user.
  * doc/examples.html, doc/notify.html: upper and lower case, format.
  * doc/*.html, man/*.?: replace; converted from markdown source.
  * socklog-check.c, socklog.c: check length of unix socket to match struct
    sockaddr.
  * pathexec_env.c, socklog-conf.c, socklog.c, tryto.c, uncat.c: whitespace.
  * socklog.c: fix compilation warning -Wpointer-sign.
  * socklog.c, man/socklog.8: add -T to drop timestamps from syslog messages
    (github/leahneukirchen).
  * buffer_put.c, byte.h, byte_chr.c, byte_copy.c, byte_cr.c, byte_diff.c,
    byte_rchr.c, byte_zero.c, str_chr.c, str_len.c: sync daemontools library
    fork with runit source.

socklog 2.1.2
Sun, 02 Nov 2025 11:41:11 +0100
  * alloc.c, alloc_re.c, buffer_0.c, byte_chr.c, byte_copy.c, byte_cr.c,
    byte_diff.c, byte_rchr.c, byte_zero.c, error.h, select.h2, wait.h,
    wait_nohang.c, wait_pid.c: sync daemontools library fork with runit
    source.
  * doc/*.html: !DOCTYPE.
  * socklog.c, tryto.c, uncat.c: use parameter name "unused" in sighandler
    functions.
  * hasflock.h1: add #include <unistd.h>.
  * alloc.h, buffer.c, buffer.h, buffer_get.c, buffer_put.c, buffer_write.c,
    byte.h, check-socklog-unix.c, pathexec_env.c, select.h1, select.h2,
    sig.c, sig.h, sig_catch.c, socklog-check.c, socklog-check.dist,
    socklog.c, tryto.c, tryto.dist, uncat.c, uncat.dist, wait.h: fix build
    with gcc15: in C23 f() means f(void), consequently signal handler
    functions now are f(int), alloc_free() and byte_zero() parameters need
    casts, buffer_unixwrite() removes "const" from parameter.
  * package/README: update copyright notice.
  * doc/configuration.html: replace links to daemontools faq with runit faq.
  * doc/*: switch all cr.yp.to http links to https.
  * doc/*, man/*, package/README: switch all smarden.org http links to https.
  * tryto.c: don't report failure if the final try of prog succeeds (thx
    Andras Korn).
  * doc/index.html: remove reference to mailing list misc@list.smarden.org.
  * doc/index.html: add reference to repository on github.com.

socklog 2.1.1
Fri, 11 Oct 2024 00:00:17 +0000
  * doc/install.html: add link to sha256sum.asc.
  * doc/install.html: change mail address.
  * doc/index.html: remove reference to socklog 2.0.x.
  * doc/*.html: remove <small>$Id$</small>.
  * debian/: remove; obsolete.
  * pathexec.h, pathexec_env.c, pathexec_run.c, seek_set.c, sgetopt.c,
    sgetopt.h, socklog-check.c, socklog-conf.c, socklog.c, subgetopt.c,
    subgetopt.h, trycpp.c, tryflock.c, trypoll.c, trysgact.c, trysgprm.c,
    tryto.c, tryulong64.c, trywaitp.c, uncat.c, x86cpuid.c: incorporate
    lib changes from runit-2.2.0 and adapt socklog programs to properly
    build with modern toolchains (thx Gentoo, Debian, Void Linux).
  * prot.c, prot.h, chkshsgr.c, tryshsgr.c, hasshsgr.h?, warn-shsgr:
    remove; obsolete.
  * Makefile, TARGETS: no longer check for "shortsetgroups" (thx Leah
    Neukirchen).
  * socklog.c: replace prot_gid() with setgroups() and prot_uid() with
    setuid().

socklog 2.1.0
Mon, 06 Mar 2006 12:59:33 +0000
  * socklog-conf, man/socklog-conf.8: create service directories
    /etc/sv/socklog-* instead of /etc/socklog/*; create ./check script
    for socklog-unix service.
  * doc/configuration.html, doc/install.html, doc/network.html,
    doc/notify.html: adapt to new path to service directories; utilize
    runit's sv program instead of runsvctrl.
  * man/socklog-check.8, man/socklog-conf.8, man/socklog.8, man/uncat.1:
    see also sv(8).

socklog 2.0.3
Sun, 26 Feb 2006 13:29:17 +0000
  * doc/index.html: minor.
  * socklog-check.c: change EDESTADDRREQ to ECONNREFUSED if defined and set
    by connect().
  * tryto: fix a race with child terminating immediatly (thx Alex Efros);
    catch SIGCHLD while waiting ktimeout seconds before sending KILL.
  * man/tryto.1: remove reference to timelimit(8) (thx Alex Efros).
  * socklog.check, socklog.dist: don't check usage output with -V as it
    differs on solaris (thx Asif Iqbal).

socklog 2.0.2
Sun, 10 Apr 2005 07:21:30 +0000
  * doc/readme.solaris.html: add -ldoor -lthread to src/conf-ld (thx Lars
    Kellogg-Stedman); update log/run script, mailing list address.
  * man/socklog.8: remove double entries in SEE ALSO.

socklog 2.0.1
Thu, 10 Feb 2005 10:05:55 +0000
  * Makefile: add linker options detected by trysocketlib.c to targets
    check-socklog-inet, check-socklog-inet (fixes test failure on solaris,
    thx Uffe Jakobsen).

socklog 2.0.0
Sun, 06 Feb 2005 19:57:04 +0000
  * socklog-check.check: tweak timing in program test.
  * tryto.c: new option -P: run prog in new session, and signal process
    group.
  * man/tryto.1: adapt; cleanup.
  * man/socklog-check.8: add hint about runit services using syslog
    facility; cleanup.
  * man/socklog-conf.8, man/socklog.8, man/uncat.1: cleanup.
  * doc/index.html: new stable release.
  * doc/network.html: typo.

socklog 1.5.0
Sat, 26 Jun 2004 11:28:33 +0000
  * socklog-check, man/socklog-check.8: new; check for the availability of
    a socklog service.
  * socklog-check.check, socklog-check.dist: new; program test.
  * byte_zero.c: new.
  * socklog.c, man/socklog.8: new option -U: don't override umask when
    creating domain socket.
  * package/check, package/compile, package/install-man, package/upgrade:
    minor cleanup.
  * Makefile, man/socklog-conf.8: adapt.

socklog 1.4.2
Tue, 23 Mar 2004 18:16:17 +0000
  * doc/index.html, doc/install.html: minor.
  * socklog.c: fix sigterm handling in ucspi mode.
  * uncat.check: tweak timing in program test.
  * Makefile: don't run target check by default.
  * package/check: new; run checks on programs.
  * package/install: run package/check.

socklog 1.4.1

src/socklog-conf.c
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <pwd.h>
#include <grp.h>
#include <stdio.h>
#include "error.h"
#include "strerr.h"
#include "open.h"
#include "buffer.h"

#if defined(__sun__) && defined(__sparc__) && defined(__unix__) && defined(__svr4__)
#define SOLARIS
#include <sys/utsname.h>
#include "scan.h"
#endif

#ifndef SOLARIS
#include <paths.h>
#endif
#ifndef _PATH_KLOG
#define _PATH_KLOG "/dev/klog"
#endif

#define USAGE1 " unix|inet|klog|ucspi acct logacct [/etc/sv] [/logdir]"
#define USAGE2 " notify acct grp [/etc/sv] [pipe]"

#define VERSION "$Id$"
#define WARNING "socklog-conf: warning: "
#define FATAL "socklog-conf: fatal: "

#define CONF_DIR "/etc/sv"
#define LOG_DIR_UNIX "/var/log/socklog"
#define LOG_DIR_INET "/var/log/socklog-inet"
#define LOG_DIR_UCSPI_TCP "/var/log/socklog-ucspi-tcp"
#define LOG_DIR_KLOG "/var/log/socklog-klog"
#define PATH_NOTIFY "/var/log/socklog/.notify"

#define CONF_UNIX 0
#define CONF_INET 1
#define CONF_UCSPI_TCP 2
#define CONF_NOTIFY 3
#define CONF_KLOG 4

const char *progname;

void usage() {
  strerr_warn3("usage: ", progname, USAGE1, 0);
  strerr_die4x(1, "usage: ", progname, USAGE2, "\n");
}

const char *dir;
const char *fn;
char buf[1024];
int fd;
buffer b;

char *user;
char *loguser;
const char *path;

struct passwd *upw, *pw;
struct group *gr;

void fail() {
  if (fn[0] == '/') 
    strerr_die4sys(111, FATAL, "unable to create ", fn, ": ");
  else
    strerr_die6sys(111, FATAL, "unable to create ", dir, "/", fn, ": ");
}

void start(const char *s) {
  fn = s;
  fd = open_trunc(fn);
  if (fd == -1) fail();
  buffer_init(&b, buffer_unixwrite, fd, buf, sizeof buf);
}

void outs(const char *s) {
  if (buffer_puts(&b, s) == -1) fail();
}

void finish(void) {
  if (buffer_flush(&b) == -1) fail();
  if (fsync(fd) == -1) fail();
  close(fd);
}

void perm(int mode) { if (chmod(fn, mode) == -1) fail(); }

void mkfile(const char *f, const char *s) {
  start(f);
  outs(s); outs("\n");
  finish();
}

void makedir(const char *s) {
  fn =s;
  if (mkdir(fn, 0750) == -1) fail();
}
void makechdir(const char *s) {
  makedir(s);
  if (chown(s, pw->pw_uid, pw->pw_gid) == -1)
    strerr_die6sys(111, FATAL, "unable to set owner of ", dir, "/", s, ": ");
}

void conf_unix() {
#ifdef SOLARIS
#if WANT_SUN_DOOR
  struct utsname u;
  unsigned long sunos_version;
  uname(&u);
  scan_ulong(u.release+strlen(u.release)-1, &sunos_version);
#endif
#endif
  makedir("socklog-unix");
  perm(0750);
  makedir("socklog-unix/log");
  perm(0755);

  makechdir(path);
  if (symlink(path, "socklog-unix/log/main") == -1)
    strerr_die4sys(111, FATAL, "unable to link ", path, ": ");
  makechdir("socklog-unix/log/main/auth");
  mkfile("socklog-unix/log/main/auth/config", "s999999\nn5\n-*\n+auth.*\n+authpriv.*");
  makechdir("socklog-unix/log/main/cron");
  mkfile("socklog-unix/log/main/cron/config", "s999999\nn5\n-*\n+cron.*");
  makechdir("socklog-unix/log/main/daemon");
  mkfile("socklog-unix/log/main/daemon/config", "s999999\nn5\n-*\n+daemon.*");
  makechdir("socklog-unix/log/main/debug");
  mkfile("socklog-unix/log/main/debug/config", "s999999\nn5\n-*\n+*.debug*");
  makechdir("socklog-unix/log/main/ftp");
  mkfile("socklog-unix/log/main/ftp/config", "s999999\nn5\n-*\n+ftp.*");
  makechdir("socklog-unix/log/main/kern");
  mkfile("socklog-unix/log/main/kern/config", "s999999\nn5\n-*\n+kern.*");
  makechdir("socklog-unix/log/main/local");
  mkfile("socklog-unix/log/main/local/config", "s999999\nn5\n-*\n+local.*");
  makechdir("socklog-unix/log/main/mail");
  mkfile("socklog-unix/log/main/mail/config", "s999999\nn5\n-*\n+mail.*");
  makechdir("socklog-unix/log/main/main");
  mkfile("socklog-unix/log/main/main/config", "s999999\nn10");
  makechdir("socklog-unix/log/main/news");
  mkfile("socklog-unix/log/main/news/config", "s999999\nn5\n-*\n+news.*");
  makechdir("socklog-unix/log/main/syslog");
  mkfile("socklog-unix/log/main/syslog/config", "s999999\nn5\n-*\n+syslog.*");
  makechdir("socklog-unix/log/main/user");
  mkfile("socklog-unix/log/main/user/config", "s999999\nn5\n-*\n+user.*");
  
  start("socklog-unix/run");
  outs("#!/bin/sh\n");
  outs("exec 2>&1\n");
#ifndef SOLARIS
  outs("exec chpst -U"); outs(user); outs(" socklog unix /dev/log\n");
#else
  outs("exec chpst -U"); outs(user); outs(" socklog sun_stream /dev/log");
#if WANT_SUN_DOOR
  if (sunos_version == 7) outs(" /etc/.syslog_door");
  if (sunos_version >= 8) outs(" /var/run/syslog_door");
#endif
  outs("\n");
#endif
  finish();
  perm(0750);

  start("socklog-unix/check");
  outs("#!/bin/sh\n");
  outs("exec 2>/dev/null\n");
  outs("exec socklog-check unix /dev/log\n");
  finish();
  perm(0755);

  start("socklog-unix/log/run");
  outs("#!/bin/sh\n");
  outs("exec chpst -u");
  outs(loguser);
  outs(" svlogd \\\n");
  outs("  main/main main/auth main/cron main/daemon main/debug main/ftp \\\n");
  outs("  main/kern main/local main/mail main/news main/syslog main/user\n");
  finish();
  perm(0750);
}

void conf_inet() {
  makedir("socklog-inet");
  perm(0750);
  makedir("socklog-inet/log");
  perm(0755);

  makechdir(path);
  if (symlink(path, "socklog-inet/log/main") == -1)
    strerr_die4sys(111, FATAL, "unable to link ", path, ": ");
  makechdir("socklog-inet/log/main/main");
  mkfile("socklog-inet/log/main/main/config", "s999999\nn10");

  start("socklog-inet/run");
  outs("#!/bin/sh\n");
  outs("exec 2>&1\n");
  outs("exec chpst -U"); outs(user); outs(" socklog inet 0 514\n");
  finish();
  perm(0750);

  start("socklog-inet/log/run");
  outs("#!/bin/sh\n");
  outs("exec chpst -u");
  outs(loguser);
  outs(" svlogd -t main/main\n");
  finish();
  perm(0750);
}

void conf_ucspi_tcp() {
  makedir("socklog-ucspi-tcp");
  perm(0750);
  makedir("socklog-ucspi-tcp/log");
  perm(0755);

  makechdir(path);
  if (symlink(path, "socklog-ucspi-tcp/log/main") == -1)
    strerr_die4sys(111, FATAL, "unable to link ", path, ": ");
  makechdir("socklog-ucspi-tcp/log/main/main");
  mkfile("socklog-ucspi-tcp/log/main/main/config", "s999999\nn10");

  start("socklog-ucspi-tcp/run");
  outs("#!/bin/sh\n");
  outs("PORT=10116\n");
  outs("exec 2>&1\n");
  outs("exec tcpsvd -vllogserver -u");
  outs(user);
  outs(" 0 \"$PORT\" socklog ucspi TCPREMOTEIP\n");
  finish();
  perm(0750);

  start("socklog-ucspi-tcp/log/run");
  outs("#!/bin/sh\n");
  outs("exec chpst -u");
  outs(loguser);
  outs(" svlogd -t main/main\n");
  finish();
  perm(0750);
}

void conf_klog() {
  makedir("socklog-klog");
  perm(0750);
  makedir("socklog-klog/log");
  perm(0755);

  makechdir(path);
  if (symlink(path, "socklog-klog/log/main") == -1)
    strerr_die4sys(111, FATAL, "unable to link ", path, ": ");
  makechdir("socklog-klog/log/main/main");
  mkfile("socklog-klog/log/main/main/config", "s999999\nn10");

  start("socklog-klog/run");
  outs("#!/bin/sh\n");
  outs("exec <"); outs(_PATH_KLOG); outs("\n");
  outs("exec 2>&1\n");
  outs("exec chpst -u");
  outs(user);
  outs(" socklog ucspi\n");
  finish();
  perm(0750);

  start("socklog-klog/log/run");
  outs("#!/bin/sh\n");
  outs("exec chpst -u");
  outs(loguser);
  outs(" svlogd -t main/main\n");
  finish();
  perm(0750);
}

void conf_notify() {
  makedir("socklog-notify");
  perm(0755);

  umask(007);
  if (mkfifo(path, 0620) == -1)
    strerr_die4sys(111, FATAL, "unable to create \"", path, "\": ");
  umask(022);
  if (chown(path, upw->pw_uid, gr->gr_gid) == -1)
    strerr_die4sys(111, FATAL, "unable to set owner of ", path, ": ");

  start("socklog-notify/run");
  outs("#!/bin/sh -e\n");
  outs("MAILTO=root\n");
  outs("PIPE="); outs(path); outs("\n\n");
  outs("if [ ! -p \"$PIPE\" ]; then mkfifo -m0620 \"$PIPE\"; chown ");
  outs(user); outs(":"); outs(loguser);
  outs(" \"$PIPE\"; fi\n");
  outs("exec <> \"$PIPE\"\n");
  outs("exec chpst -u");
  outs(user);
  outs(" uncat -s49999 -t180 \\\n");
  outs("  env MAILUSER=log MAILNAME='socklog notify' \\\n");
  outs("    mail -s socklog-notify $MAILTO\n");
  finish();
  perm(0750);
}

int main(int argc, char **argv) {
  int mode =0;

  progname =*argv++;
  umask(022);

  if (! *argv) usage();
  switch (**argv) {
  case 'u':
    if (! *(++*argv)) usage();
    switch (**argv) {
    case 'n':
      mode =CONF_UNIX;
      break;
    case 'c':
      mode =CONF_UCSPI_TCP;
      break;
    default:
      usage();
    }
    break;
  case 'i':
    mode =CONF_INET;
    break;
  case 'n':
    mode =CONF_NOTIFY;
    break;
  case 'k':
    mode =CONF_KLOG;
    break;
  case '-':
    if ((*argv)[1] && (*argv)[1] == 'V') {
      strerr_warn2(VERSION, "\n", 0);
    }
  default:
    usage();
  }
  argv++;

  user =*argv++;
  if (!user) usage();
  loguser =*argv++;
  if (!loguser) usage();
  dir =*argv++;
  if (!dir) {
    dir =CONF_DIR;
  } else {
    if (dir[0] != '/') usage();
    if (*argv) {
      path =*argv;
      if (path[0] != '/') usage();
    }
  }

  upw =getpwnam(user);
  if (!upw) strerr_die3x(111, FATAL, "unknown account ", user);
  if (mode != CONF_NOTIFY) {
    pw =getpwnam(loguser);
    if (!pw) strerr_die3x(111, FATAL, "unknown account ", loguser);
  } else {
    gr =getgrnam(loguser);
    if (!gr) strerr_die3x(111, FATAL, "unknown group ", loguser);
  }

  if (chdir(dir) == -1) {
    if (errno != error_noent)
      strerr_die4sys(111, FATAL, "unable to switch to ", dir, ": ");

    if (mkdir(dir, 0700) == -1)
      strerr_die4sys(111, FATAL, "unable to create ", dir, ": ");
    if (chmod(dir, 0750) == -1)
      strerr_die4sys(111, FATAL, "unable to set mode of ", dir, ": ");
    if (chdir(dir) == -1)
      strerr_die4sys(111, FATAL, "unable to switch to ", dir, ": ");
  }
  
  switch(mode) {
  case CONF_UNIX:
    if (!path) path =LOG_DIR_UNIX;
    conf_unix();
    break;
  case CONF_INET:
    if (!path) path =LOG_DIR_INET;
    conf_inet();
    break;
  case CONF_UCSPI_TCP:
    if (!path) path =LOG_DIR_UCSPI_TCP;
    conf_ucspi_tcp();
    break;
  case CONF_KLOG:
    if (!path) path =LOG_DIR_KLOG;
    conf_klog();
    break;
  case CONF_NOTIFY:
    if (!path) path =PATH_NOTIFY;
    conf_notify();
    break;
  }
  return(0);
}
